src/audio/codec_adapter/codec/waves.c

```diff
@@ -0,0 +1,754 @@
+static int waves_effect_buffers(struct comp_dev *dev)
+{
+	struct codec_data *codec = comp_get_codec(dev);
+	struct waves_codec_data *waves_codec = codec->private;
+	MaxxStatus_t status;
+
+	comp_dbg(dev, "waves_effect_buffers() start");
+
+	/* memmory for response
+	 * SOF does not support get requests from codec adapter right now
+	 * response will be stored in internal buffer to be dumped into logs
+	 */
+	status = MaxxEffect_GetMessageMaxSize(waves_codec->effect, &waves_codec->request_max_bytes,
+					      &waves_codec->response_max_bytes);
+
+	if (status) {
+		comp_err(dev, "waves_effect_buffers() MaxxEffect_GetMessageMaxSize() error %d",
+			 status);
+		return -EINVAL;
+	}
+
+	waves_codec->response = codec_allocate_memory(dev, waves_codec->response_max_bytes, 16);
+	if (!waves_codec->response) {
+		comp_err(dev, "waves_effect_buffers() memory for response");
+		return -ENOMEM;
+	}
+
+	/* memory for input/output buffers */
+	waves_codec->i_buffer = codec_allocate_memory(dev, waves_codec->buffer_bytes, 16);
+	if (!waves_codec->i_buffer) {
+		codec_free_memory(dev, waves_codec->response);
+		comp_err(dev, "waves_effect_buffers() allocate memory for i_buffer");
+		return -ENOMEM;
+	}
+
+	waves_codec->o_buffer = codec_allocate_memory(dev, waves_codec->buffer_bytes, 16);
+	if (!waves_codec->o_buffer) {
+		codec_free_memory(dev, waves_codec->response);
+		codec_free_memory(dev, waves_codec->i_buffer);
+		comp_err(dev, "waves_effect_buffers() allocate memory for o_buffer");
+		return -ENOMEM;
+	}
+
+	comp_info(dev, "waves_effect_buffers() size response %d, i_buffer %d, o_buffer %d",
+		  waves_codec->response_max_bytes, waves_codec->buffer_bytes,
+		  waves_codec->buffer_bytes);
+
+	codec->cpd.in_buff = waves_codec->i_buffer;
+	codec->cpd.in_buff_size = waves_codec->buffer_bytes;
+	codec->cpd.out_buff = waves_codec->o_buffer;
+	codec->cpd.out_buff_size = waves_codec->buffer_bytes;
+
+	comp_dbg(dev, "waves_effect_buffers() done");
+	return 0;
+}
+
+#if CONFIG_TRACE
+/* traces uint32_t array of len elements into debug messages */
+static void trace_array(const struct comp_dev *dev, const uint32_t *arr, uint32_t len)
+{
+	uint32_t i;
+
+	for (i = 0; i < len; i++)
+		comp_dbg(dev, "trace_array() data[%03d]:0x%08x", i, *(arr + i));
+}
+
+#define REV(N) (\
+	{typeof(N) n = (N);\
+	((n & 0x000000FF) << 24) |\
+	((n & 0x0000FF00) << 8) |\
+	((n & 0x00FF0000) >> 8) |\
+	((n & 0xFF000000) >> 24); })
+
+#define DUMP_REVISION_ITERATION(ptr, len) do {\
+	typeof(ptr) p = ptr;\
+	typeof(len) l = len;\
+	if (l >= 4) {\
+		comp_info(dev, "%08x%08x%08x%08x", REV(p[0]), REV(p[1]), REV(p[2]), REV(p[3]));\
+		l -= 4; p += 4;\
+	} else if (l == 3) {\
+		comp_info(dev, "%08x%08x%08x", REV(p[0]), REV(p[1]), REV(p[2]));\
+		l -= 3; p += 3;\
+	} else if (l == 2) {\
+		comp_info(dev, "%08x%08x", REV(p[0]), REV(p[1]));\
+		l -= 2; p += 2;\
+	} else if (l == 1) {\
+		comp_info(dev, "%08x", REV(p[0]));\
+		l--; p++;\
+	} \
+	ptr = p; len = l;\
+} while (0)
+#endif
```

Comment:
Can we make this generic code and place it in debug.h - very useful for debugging and dumping hex data.

Reply:
just move this definition to debug.h as is?

Reply:
Yes, move it to a generic place like debug.h but not as it is. 
It looks like it is dumping the first 4 bytes in hexa stored in a pointer right?

Lets try to call this `dump_hex(ptr, len)`. Also, what's `REV`?

Reply:
arguments are ptr - pointer to a buffer with 32-bit words, len - number of 32-bit words in the buffer
macro dumps up-to 4 first 32-bit words from ptr depending on len.
each word is endianness-swapped. This is what REV does, I was suggested to rename it to BYTESWAPP in the upcoming revision.

Also, macro increments the pointer by the number of words dumped and decrements len by the same amount.
Will it be ok for a "generic" macro to do so?

Reply:
hi @dbaluta, @lgirdwood
done, please check

still not sure this should be part of a framework

Reply:
@stolx please switch your code to use that macro

Reply:
done
Also, I changed `dump_hex` macro implementation, now it only modifies idx by a number of words printed